cloud/foreman/foreman.go

```
@@ -37,189 +32,223 @@ type clientImpl struct {
 
-		if m.PrivateIP != "" && m.PublicKey != "" {
-			minionIPToPublicKey[m.PrivateIP] = m.PublicKey
+		if !ok {
+			stop := make(chan struct{})
+			minionChans[m.CloudID] = stop
+			go newMinion(conn, m.CloudID, stop)
 		}
+
+		seen[m.CloudID] = struct{}{}
 	}
 
-	// Assign all of the minions their new configs
-	forEachMinion(func(m *minion) {
-		if !m.connected {
-			return
+	for id := range minionChans {
+		if _, ok := seen[id]; !ok {
+			log.Debugf("Stop minion at %s", id)
+			close(minionChans[id])
+			delete(minionChans, id)
 		}
+	}
+}
 
-		newConfig := pb.MinionConfig{
-			FloatingIP:          m.machine.FloatingIP,
-			PrivateIP:           m.machine.PrivateIP,
-			Blueprint:           blueprint,
-			Provider:            string(m.machine.Provider),
-			Size:                m.machine.Size,
-			Region:              m.machine.Region,
-			EtcdMembers:         etcdIPs,
-			AuthorizedKeys:      m.machine.SSHKeys,
-			MinionIPToPublicKey: minionIPToPublicKey,
+var newMinion = newMinionImpl
+
+func newMinionImpl(conn db.Conn, cloudID string, stop chan struct{}) {
+	log.Debugf("Start minion at %s", cloudID)
+	// Threads that aren't currently connected to their minion should run more often.
+	frequentTrigger := conn.TriggerTick(5, db.BlueprintTable, db.MachineTable)
+	trigger := conn.TriggerTick(60, db.BlueprintTable, db.MachineTable)
+	defer frequentTrigger.Stop()
+	defer trigger.Stop()
+
+	var connected bool
+	for {
+		select {
+		case <-stop:
+		case <-frequentTrigger.C:
+			if connected {
+				continue
+			}
+		case <-trigger.C:
 		}
 
-		if reflect.DeepEqual(newConfig, m.config) {
+		// In a race between a closed stop and a trigger, choose stop.
+		select {
+		case <-stop:
 			return
+		default:
 		}
 
-		if err := m.client.setMinion(newConfig); err != nil {
+		currConfig, connected := runOnce(conn, cloudID)
+		setMinionStatus(cloudID, currConfig, connected)
+	}
+}
+
+func runOnce(conn db.Conn, cloudID string) (currConfig pb.MinionConfig, connected bool) {
```

I think you could reorganize this to be just function to be just a bit simpler.   Specifically, I don't much like that we need to do two separate database transactions, and thus have two separate opportunities for the world to change out from under us as we're running.

What if you first thing you did a select to get the list of the machines.  Then Given that list of machines, you search through it to find the machine with my cloudID (which will give you the public IP).  And then later in makeConcifg, you can pass in the list of machines you grabbed earlier, and the machine that represents yourself.

---

That would also simplify makeConfig so that it doesn't need to return an error.  You can just deal with the "lack of self" error once at the top of runOnce()

---

Ah, I like that. Thanks!